File: lua/test/photo_station_entity_test.lua
-- PhotoStation entity test

local json = require("dkjson")
local vs = require("utility.struct.struct")
local sdk = require("railway-station-photos_sdk")
local helpers = require("core.helpers")
local runner = require("test.runner")

local _test_dir = debug.getinfo(1, "S").source:match("^@(.+/)")  or "./"

describe("PhotoStationEntity", function()
  it("should create instance", function()
    local testsdk = sdk.test(nil, nil)
    local ent = testsdk:PhotoStation(nil)
    assert.is_not_nil(ent)
  end)

  it("should run basic flow", function()
    local setup = photo_station_basic_setup(nil)
    -- Per-op sdk-test-control.json skip.
    local _live = setup.live or false
    for _, _op in ipairs({"list", "load"}) do
      local _should_skip, _reason = runner.is_control_skipped("entityOp", "photo_station." .. _op, _live and "live" or "unit")
      if _should_skip then
        pending(_reason or "skipped via sdk-test-control.json")
        return
      end
    end
    -- The basic flow consumes synthetic IDs from the fixture. In live mode
    -- without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only then
      pending("live entity test uses synthetic IDs from fixture — set RAILWAYSTATIONPHOTOS_TEST_PHOTO_STATION_ENTID JSON to run live")
      return
    end
    local client = setup.client

    -- Bootstrap entity data from existing test data.
    local photo_station_ref01_data_raw = vs.items(helpers.to_map(
      vs.getpath(setup.data, "existing.photo_station")))
    local photo_station_ref01_data = nil
    if #photo_station_ref01_data_raw > 0 then
      photo_station_ref01_data = helpers.to_map(photo_station_ref01_data_raw[1][2])
    end

    -- LIST
    local photo_station_ref01_ent = client:PhotoStation(nil)
    local photo_station_ref01_match = {}

    local photo_station_ref01_list_result, err = photo_station_ref01_ent:list(photo_station_ref01_match, nil)
    assert.is_nil(err)
    assert.is_table(photo_station_ref01_list_result)

    -- LOAD
    local photo_station_ref01_match_dt0 = {}
    local photo_station_ref01_data_dt0_loaded, err = photo_station_ref01_ent:load(photo_station_ref01_match_dt0, nil)
    assert.is_nil(err)
    assert.is_not_nil(photo_station_ref01_data_dt0_loaded)

  end)
end)

function photo_station_basic_setup(extra)
  runner.load_env_local()

  local entity_data_file = _test_dir .. "../../.sdk/test/entity/photo_station/PhotoStationTestData.json"
  local f = io.open(entity_data_file, "r")
  if f == nil then
    error("failed to read photo_station test data: " .. entity_data_file)
  end
  local entity_data_source = f:read("*a")
  f:close()

  local entity_data = json.decode(entity_data_source)

  local options = {}
  options["entity"] = entity_data["existing"]

  local client = sdk.test(options, extra)

  -- Generate idmap via transform.
  local idmap = vs.transform(
    { "photo_station01", "photo_station02", "photo_station03", "photo_station_by_id01", "photo_station_by_id02", "photo_station_by_id03", "photo_stations_by_country01", "photo_stations_by_country02", "photo_stations_by_country03", "photo_stations_by_photographer01", "photo_stations_by_photographer02", "photo_stations_by_photographer03" },
    {
      ["`$PACK`"] = { "", {
        ["`$KEY`"] = "`$COPY`",
        ["`$VAL`"] = { "`$FORMAT`", "upper", "`$COPY`" },
      }},
    }
  )

  -- Detect ENTID env override before envOverride consumes it. When live
  -- mode is on without a real override, the basic test runs against synthetic
  -- IDs from the fixture and 4xx's. Surface this so the test can skip.
  local entid_env_raw = os.getenv("RAILWAYSTATIONPHOTOS_TEST_PHOTO_STATION_ENTID")
  local idmap_overridden = entid_env_raw ~= nil and entid_env_raw:match("^%s*{") ~= nil

  local env = runner.env_override({
    ["RAILWAYSTATIONPHOTOS_TEST_PHOTO_STATION_ENTID"] = idmap,
    ["RAILWAYSTATIONPHOTOS_TEST_LIVE"] = "FALSE",
    ["RAILWAYSTATIONPHOTOS_TEST_EXPLAIN"] = "FALSE",
    ["RAILWAYSTATIONPHOTOS_APIKEY"] = "NONE",
  })

  local idmap_resolved = helpers.to_map(
    env["RAILWAYSTATIONPHOTOS_TEST_PHOTO_STATION_ENTID"])
  if idmap_resolved == nil then
    idmap_resolved = helpers.to_map(idmap)
  end

  if env["RAILWAYSTATIONPHOTOS_TEST_LIVE"] == "TRUE" then
    local merged_opts = vs.merge({
      {
        apikey = env["RAILWAYSTATIONPHOTOS_APIKEY"],
      },
      extra or {},
    })
    client = sdk.new(helpers.to_map(merged_opts))
  end

  local live = env["RAILWAYSTATIONPHOTOS_TEST_LIVE"] == "TRUE"
  return {
    client = client,
    data = entity_data,
    idmap = idmap_resolved,
    env = env,
    explain = env["RAILWAYSTATIONPHOTOS_TEST_EXPLAIN"] == "TRUE",
    live = live,
    synthetic_only = live and not idmap_overridden,
    now = os.time() * 1000,
  }
end

File: rb/test/photo_station_entity_test.rb
# PhotoStation entity test

require "minitest/autorun"
require "json"
require_relative "../RailwayStationPhotos_sdk"
require_relative "runner"

class PhotoStationEntityTest < Minitest::Test
  def test_create_instance
    testsdk = RailwayStationPhotosSDK.test(nil, nil)
    ent = testsdk.PhotoStation(nil)
    assert !ent.nil?
  end

  def test_basic_flow
    setup = photo_station_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["list", "load"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "photo_station." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set RAILWAYSTATIONPHOTOS_TEST_PHOTO_STATION_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # Bootstrap entity data from existing test data.
    photo_station_ref01_data_raw = Vs.items(Helpers.to_map(
      Vs.getpath(setup[:data], "existing.photo_station")))
    photo_station_ref01_data = nil
    if photo_station_ref01_data_raw.length > 0
      photo_station_ref01_data = Helpers.to_map(photo_station_ref01_data_raw[0][1])
    end

    # LIST
    photo_station_ref01_ent = client.PhotoStation(nil)
    photo_station_ref01_match = {}

    photo_station_ref01_list_result = photo_station_ref01_ent.list(photo_station_ref01_match, nil)
    assert photo_station_ref01_list_result.is_a?(Array)

    # LOAD
    photo_station_ref01_match_dt0 = {}
    photo_station_ref01_data_dt0_loaded = photo_station_ref01_ent.load(photo_station_ref01_match_dt0, nil)
    assert !photo_station_ref01_data_dt0_loaded.nil?

  end
end

def photo_station_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "photo_station", "PhotoStationTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = RailwayStationPhotosSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["photo_station01", "photo_station02", "photo_station03", "photo_station_by_id01", "photo_station_by_id02", "photo_station_by_id03", "photo_stations_by_country01", "photo_stations_by_country02", "photo_stations_by_country03", "photo_stations_by_photographer01", "photo_stations_by_photographer02", "photo_stations_by_photographer03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["RAILWAYSTATIONPHOTOS_TEST_PHOTO_STATION_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "RAILWAYSTATIONPHOTOS_TEST_PHOTO_STATION_ENTID" => idmap,
    "RAILWAYSTATIONPHOTOS_TEST_LIVE" => "FALSE",
    "RAILWAYSTATIONPHOTOS_TEST_EXPLAIN" => "FALSE",
  })

  idmap_resolved = Helpers.to_map(
    env["RAILWAYSTATIONPHOTOS_TEST_PHOTO_STATION_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["RAILWAYSTATIONPHOTOS_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
      },
      extra || {},
    ])
    client = RailwayStationPhotosSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["RAILWAYSTATIONPHOTOS_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["RAILWAYSTATIONPHOTOS_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end
